### ISSUE TITLE
How to have a multiline value

### ISSUE THREAD

**Issue by Youssef1313** [editorconfig/editorconfig#481]
See https://github.com/editorconfig/specification/issues/35

> Per the current specification, I don't see how (if possible) to include a key-value pair whose value is a multiline string.
> 
> If it's not currently possible, is it possible to update the spec to have a way for that?

**Comment by cxw42** [comment]
Thanks for using EC!  Could you give us a sample use case?  What key are you using, and what would a sample multiline value be?

**Comment by Youssef1313** [comment]
@cxw42 It will be usefuly for the C# compiler. (Note: I'm not a Microsoft employee, but a regular contributor to Roslyn)

In short, the C# compiler currently uses .editorconfig files to parse analyzer options. Most of the time these are single-line key-value pairs, e.g, `dotnet_diagnostic.CA1822.severity = warning`.

However, MSBuild properties (xml-based) can be passed to the compiler, for that, there is a [build task](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs) that converts the "compiler visible" MSBuild properties to an .editorconfig during compile-time.

MSBuild property values can be multiline, and here comes the problem.

See issue: https://github.com/dotnet/roslyn/issues/43970

**Comment by cxw42** [comment]
Thanks for the additional information!  I see the linked issue gives this example (I added `{{` `}}` around the desired property value):
```
is_global = true
msbuild_property.PropWithNewLine = {{this
is 
a
valid = property}}
```

The linked issue talks about "global analyzer configs" --- would the multiline properties be consumed by editors or other tools?

This seems to me like a very specialized use-case that is probably outside the scope of the specification.  However, others may certainly have different ideas!

**Comment by Youssef1313** [comment]
> would the multiline properties be consumed by editors or other tools?

would the multiline properties be consumed by editors or other tools?

Yes, this will be consumed by the C#/VB compiler (ships in Visual Studio).

The current behavior is that [this task](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/MSBuildTask/GenerateMSBuildEditorConfig.cs) generates a broken .editorconfig file.

**Comment by chrisjohnson00** [comment]
Here is another use case, I want to enforce 120 character line lengths, but the editor config file itself fails the validation due to one of the keys and values being long, 144 chars:

```
dotnet_naming_symbols.non_field_members.applicable_accessibilities = public, internal, private, protected, protected_internal, private_protected
```

**Comment by cxw42** [comment]
@chrisjohnson00 that looks like noncompliant behaviour.  The [spec](https://spec.editorconfig.org/#id9) says:
> Cores must accept keys and values with lengths up to and including 1024 and 4096 characters respectively.

I suggest opening issues against whatever program is giving you the validation error on the editorconfig file.

**Comment by CodePagesNet** [comment]
Yes, the ability for a config value to span multiple lines would be helpful, e.g. for a property such as the cpp_include_cleanup_excluded_files custom Visual Studio property.

**Comment by RampantDespair** [comment]
Was this ever implemented?

**Comment by cxw42** [comment]
Not that I am aware of.

**Comment by RampantDespair** [comment]
@cxw42 
I see, so in the event where I would like to contribute, can you point me to the repository that holds the logic pertaining to this?
I can't seem to find it sadly.

**Comment by cxw42** [comment]
@RampantDespair I really appreciate your asking, and your willingness to help!  Unfortunately, this is not ready for implementation.  If you'd like to make a prototype, you'll need to edit whatever core and plugin you're using.  E.g., I use Vim, so I would edit https://github.com/editorconfig/editorconfig-vim if I were going to make a proof of concept.

If you want to prototype only on the core side, you can use https://github.com/editorconfig/editorconfig-core-c or https://github.com/editorconfig/editorconfig-core-py, for example.

Multiline-property support can't be rolled out without making a breaking change, unfortunately, to both cores and plugins.  To fully support multiline properties would take:
* A spec-change proposal
* A vote of the board
* Updating all the cores
* Updating all the plugins

It's not a small task :( .

I opened a proposal a while back at #482 that would use TOML as the base format, and thereby give us multiline values for free.  Feel free to comment there if you wish!  That proposal is not in use anywhere currently, as far as I know.

**Comment by RampantDespair** [comment]
@cxw42 
Wow. Your request dates back to 2023, and is still not approved, that's really unfortunate. :(
In all fairness multi-line support should only be the start of this change.

Because my main problem, and what prompted me to research whether multi-line was even possible is this setting:
`cpp_include_cleanup_alternate_files`
Currently that setting is sitting at >25,000 characters of length, and I literally had to create a script to automatically generate it.

I say this should only be the start, because another issue I have with it, is that it doesn't support inheritance.
So since I use libraries with barrel imports (convenience headers as some call it), this forces me to go down through each child of that header and add everything to not get pestered with warnings about explicit includes for vendor libraries.

As an example:
A.h includes B.h includes C.h,
this results in:
`A.h:B.h,A.h:C.h,B.h:C.h`

So basically my idea, which to me is the only solution, would be to support trees obviously.
But to support trees we'd need multi-line support ideally.

So hopefully your multi-line support is accepted in 2026, that way proposing my tree idea will be better received I'd imagine.

EDIT: Even wildcards would be amazing and probably another great addition, if A.h is included consider A/**/* as included as well.

**Comment by cxw42** [comment]
> if A.h is included consider A/**/* as included as well.

You mean in the `cpp_include_cleanup_alternate_files` value specifically, right?  If so, feel free to open that request with Microsoft.  That property is not standardized by EditorConfig, so MS could add that wildcard support if they wanted to with no issues from EC.

**Comment by RampantDespair** [comment]
@cxw42 
I've opened a feature request with microsoft, thank you for pointing me in the right direction.
